std-bits/base/src/main/java/org/enso/base/cache/LRUCache.java

```diff
@@ -230,30 +261,40 @@ private long getTotalCacheSize() {
     return cache.values().stream().collect(Collectors.summingLong(e -> e.size()));
   }
 
-  private long getMaxFileSize() {
-    return cacheTestParameters.getMaxFileSizeOverrideTestOnly().orElse(maxFileSize);
+  /**
+   * Calculate the max total cache size, using the current limit but also constraining the result to
+   * the upper bound.
+   */
+  public long getMaxTotalCacheSize(long currentlyUsed) {
+    var totalCacheSize =
+        switch (settings.getTotalCacheLimit()) {
+          case TotalCacheLimit.Bytes bytes -> bytes.bytes();
+          case TotalCacheLimit.Percentage percentage -> {
+            long usableSpace = diskSpaceGetter.get() + currentlyUsed;
+            yield (long) (percentage.percentage() * usableSpace);
+          }
+        };
+    return Long.min(MAX_TOTAL_CACHE_SIZE_FREE_SPACE_UPPER_BOUND, totalCacheSize);
```

Comment:
A few edge cases that I think need to be handled here:
1. in Bytes mode we completely skip checking the available disk space. What if our cache limit is set to 1000MB but the device has only 100MB of available space? This code currently returns 1000MB as the max total cache size. So the cache logic will happily perform writes for up to 1GB, exceeding the available 100MB. This will likely mean that the first cache write operation that happens when we run out of disk space will probably fail with some `IOException: Out of storage space`. But we have no logic to handle this. Running out of storage space for a cache should not mean that a HTTP request fails - from user's perspective it has nothing to do with disk storage, so the details of cache handling should be opaque to the user and not cause errors. But instead of catching this kind of error, we should be pro-actively ensuring that our cache writes do not exceed the available disk space.
2. moreover, I don't think it is good practice for a _cache_ to take up the last bytes available on the disk. As that may cause other applications to stop working properly. This is maybe less important but it feels like to be a good citizen of the OS, an optional cache should not use up the last resources. Thus I was proposing to keep a margin that is never exceeded. E.g. if the available disk space is less than 100MB (or 1GB, up to us), then regardless of total cache size set, we probably should not take up these last few bytes.

(2) is optional, but (1) is I think really needed to make this cache usable - running out of disk space should not be a cause for failing HTTP requests. But since we need to ensure we don't make the cache larger than available disk space, we probably could additionally add a little margin of ~100MB to satisfy (2) as well.

Reply:
When working on either 1 or 2, we need to take into account the `ENSO_LIB_HTTP_CACHE_MAX_FILE_SIZE` - as the cache can grow by at most this amount. So we need to ensure that before downloading any file to a cache location, we have at least this much free disk space.

Of course there's always the possibility of edge cases like other applications writing to disk at the same time as us and thus running out of space even if we had all the necessary margins. But IMHO such cases are probably ok to ignore for now. But keeping the margins is really rather easy so IMO it would be worth to do it. For the other cases - we may want to intercept the `IOException` and wrap it into some more Enso-friendly error, e.g. `Illegal_State.Error` telling the user that we ran out of disk space when trying to cache the file and that they may try to disable caches for this request or just re-run the node to try again.

Reply:
Instead of a fixed margin, I capped the total allowed size at 90% (`MAX_PERCENTAGE`) of the free disk space, regardless of whether the limit was specified in MB or as a percentage. I think this provides the protection we need.

For the max file size -- previously, it was clearing enough space for the download if there was a content-length, but not otherwise. Now, if there is no content-length, it clears out space for the largest allowed file. This is likely to be too much, but this case is probably a lot less common.

As for running out of space anyway, because of other applications writing to disk (or for any other reason) -- there is already logic to handle that case. If it encounters an `IOException` `makeRequestAndCache`, it removes any partially-downloaded file, and re-issues the request without caching. (I don't attach a warning in this case, because this is very likely to be transient.)

Reply:
Ah sorry, I did not notice the 90% cap. Okay, I think that is good indeed.

Looks great then.